### PR TITLE
Upgrade `instrumentation_library_metrics` to `scope_metrics`

### LINKED ISSUE #1492: [Metrics] Upgrade `instrumentation_library_metrics` to `scope_metrics`
**Is your feature request related to a problem?**

According to https://github.com/open-telemetry/opentelemetry-proto/blob/v0.17.0/opentelemetry/proto/metrics/v1/metrics.proto ,
`instrumentation_library_metrics` is deprecated and removed in https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto , we should upgrade to use `scope_metrics` instead, just like traces and logs.

**Describe the solution you'd like**

+ Change the name of `instrumentation_info_metric_data_` in `ResourceMetrics` and `instrumentation_info_metrics` in `MetricCollector::Collect`
+ `PopulateResourceMetrics` to `scope_metrics` in OTLP exporters
+ Rename `InstrumentationInfoMetrics` to `ScopeMetrics`
+ Rename `instrumentation_library_` to `scope_`
+ Rename `instrumentation_metrics` to `scope_metrics`

### LINKED ISSUE #1506: [Logs] Upgrade `instrumentation_library` to  `scope` in SDK and all member names of logs
Just like #1492 , rename `instrumentation_library` in logs to `scope`

### PR DESCRIPTION
**Pull request by owent**
Signed-off-by: owentou <owentou@tencent.com>

Fixes #1492
Fixed #1506 

## Changes

+ Change the name of `instrumentation_info_metric_data_` in `ResourceMetrics` and `instrumentation_info_metrics` in `MetricCollector::Collect`
+ `PopulateResourceMetrics` to `scope_metrics` in OTLP exporters
+ Rename `InstrumentationInfoMetrics` to `ScopeMetrics`
+ Rename `instrumentation_library_` to `scope_`
+ Rename `instrumentation_metrics` to `scope_metrics`
+ Rename `InstrumentationLibrary` to `InstrumentationScope`
+ Rename `GetInstrumentationLibrary` to  `GetInstrumentationScope`
+ Rename `SetInstrumentationLibrary` to  `SetInstrumentationScope`
+ Rename `printInstrumentationLibrary` to `printInstrumentationScope`
+ Rename `instrumentation_library_` to `instrumentation_scope_`
+ Rename `instrumentation_library` to `instrumentation_library`
+ Move instrumentationlibrary/instrumentation_library.h to instrumentationscope/instrumentation_scope.h
+ Add deprecated message for legacy `GetInstrumentationLibrary` and `SetInstrumentationLibrary`

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by lalitb** [comment]
Thanks for the PR. Looks good in general. One suggestion - do you think we should introduce `opentelemetry::sdk::instrumentationScope::InstrumentationScope` class and start using that for metrics (and probably also for logs). We can later deprecate "InstrumentationLibrary" once we remove it from traces. This way, we also add field `attributes` in InstrumentationScope as introduced here - https://github.com/open-telemetry/opentelemetry-specification/pull/2579

**Comment by owent** [comment]
> Thanks for the PR. Looks good in general. One suggestion - do you think we should introduce `opentelemetry::sdk::instrumentationScope::InstrumentationScope` class and start using that for metrics (and probably also for logs). We can later deprecate "InstrumentationLibrary" once we remove it from traces. This way, we also add field `attributes` in InstrumentationScope as introduced here - [open-telemetry/opentelemetry-specification#2579](https://github.com/open-telemetry/opentelemetry-specification/pull/2579). Should be fine to do this in separate or current PR.

Agree, do you think if we can rename `InstrumentationLibrary` to `InstrumentationScope` and add a alias with `using InstrumentationLibrary = InstrumentationScope`  to keep compatibility for one or two versions? It's a break change to stable trace APIs if we just change the name.

**Comment by lalitb** [comment]
> > Thanks for the PR. Looks good in general. One suggestion - do you think we should introduce `opentelemetry::sdk::instrumentationScope::InstrumentationScope` class and start using that for metrics (and probably also for logs). We can later deprecate "InstrumentationLibrary" once we remove it from traces. This way, we also add field `attributes` in InstrumentationScope as introduced here - [open-telemetry/opentelemetry-specification#2579](https://github.com/open-telemetry/opentelemetry-specification/pull/2579). Should be fine to do this in separate or current PR.
> 
> Agree, do you think if we can rename `InstrumentationLibrary` to `InstrumentationScope` and add a alias with `using InstrumentationLibrary = InstrumentationScope` to keep compatibility for one or two versions? It's a break change to stable trace APIs if we just change the name.

Yes, that's a good idea. I was actually thinking to keep both, but your suggestion seems better, as cleanup would be easy.

**Comment by owent** [comment]
> > > Thanks for the PR. Looks good in general. One suggestion - do you think we should introduce `opentelemetry::sdk::instrumentationScope::InstrumentationScope` class and start using that for metrics (and probably also for logs). We can later deprecate "InstrumentationLibrary" once we remove it from traces. This way, we also add field `attributes` in InstrumentationScope as introduced here - [open-telemetry/opentelemetry-specification#2579](https://github.com/open-telemetry/opentelemetry-specification/pull/2579). Should be fine to do this in separate or current PR.
> > 
> > 
> > Agree, do you think if we can rename `InstrumentationLibrary` to `InstrumentationScope` and add a alias with `using InstrumentationLibrary = InstrumentationScope` to keep compatibility for one or two versions? It's a break change to stable trace APIs if we just change the name.
> 
> Yes, that's a good idea. I was actually thinking to keep both, but your suggestion seems better, as cleanup would be easy.

Done

File: exporters/otlp/include/opentelemetry/exporters/otlp/otlp_log_recordable.h
Comment (lalitb) [inline exporters/otlp/include/opentelemetry/exporters/otlp/otlp_log_recordable.h:96]:
```suggestion
   * @param instrumentation_scope the instrumentation scope to set
```

File: sdk/src/trace/tracer_provider.cc
Comment (lalitb) [inline sdk/src/trace/tracer_provider.cc:69]: Just a side note (not related to this PR) - Also need to support [attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#get-a-tracer) added to InstrumentationScope as part of this PR - [Introduce Instrumentation Scope Attributes](https://github.com/open-telemetry/opentelemetry-specification/pull/2579). Will create a separate github issue for this.

File: exporters/otlp/src/otlp_metric_utils.cc
Comment (lalitb) [inline exporters/otlp/src/otlp_metric_utils.cc:217]: while you are here, can you also rename this method to `PopulateInstrumentInfoMetrics`? As this method populates instrument's information, and not related to instrumentation info/scope:)

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (lalitb) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:221]: Do we need this method `SetInstrumentationLibrary` - as it is already defined it `sdk::logs::Recordable`?

Comment (owent) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h]: Thanks, and this can be removed.

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (lalitb) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:206]: * @param instrumentation_scope the instrumentation scope to set

Comment (lalitb) [inline exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:206]:
```suggestion
   * @param instrumentation_scope the instrumentation scope to set
```

File: exporters/otlp/src/otlp_metric_utils.cc
Comment (owent) [reply exporters/otlp/src/otlp_metric_utils.cc]: Done

File: sdk/src/trace/tracer_provider.cc
Comment (owent) [reply sdk/src/trace/tracer_provider.cc:69]: OK

File: exporters/otlp/include/opentelemetry/exporters/otlp/otlp_log_recordable.h
Comment (owent) [reply exporters/otlp/include/opentelemetry/exporters/otlp/otlp_log_recordable.h]: Thanks and fixed.

File: exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h
Comment (owent) [reply exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:208]: Thanks and fixed.

**Review by lalitb — APPROVED** [review]
LGTM once all the comments are resolved.

File: sdk/src/trace/tracer_provider.cc
Comment (owent) [reply sdk/src/trace/tracer_provider.cc:69]: Can we add attributes in another PR.This PR just change the field name and use the new protobuf field to upload OTLP datas. So we can support the new system first.

Comment (lalitb) [reply sdk/src/trace/tracer_provider.cc:69]: Yeah attributes should be separate PR. Will resolve this comment.
Thanks

**Comment by lalitb** [comment]
This looks good to merge, but being bigger PR, we can wait for @esigo or @ThomsonTan to review it.

**Comment by owent** [comment]
> This looks good to merge, but being bigger PR, we can wait for @esigo or @ThomsonTan to review it.

Thanks

**Review by esigo — APPROVED** [review]
LGTM
Thanks @owent for the PR :)

**Comment by lalitb** [comment]
@owent  - Need to rebase this to merge :)

**Comment by owent** [comment]
> @owent - Need to rebase this to merge :)

Thanks and done.

**Pull request merged**
